go/vt/vttablet/onlineddl/executor.go

```diff
@@ -334,85 +335,86 @@ func (e *Executor) triggerNextCheckInterval() {
 // restrictions on which migrations exactly are allowed such concurrency.
-func (e *Executor) allowConcurrentMigration(onlineDDL *schema.OnlineDDL) bool {
+func (e *Executor) allowConcurrentMigration(onlineDDL *schema.OnlineDDL) (action sqlparser.DDLAction, allowConcurrent bool) {
 	if !onlineDDL.StrategySetting().IsAllowConcurrent() {
-		return false
+		return action, false
 	}
 
-	action, err := onlineDDL.GetAction()
+	var err error
+	action, err = onlineDDL.GetAction()
 	if err != nil {
-		return false
+		return action, false
 	}
 	switch action {
 	case sqlparser.CreateDDLAction, sqlparser.DropDDLAction:
 		// CREATE TABLE, DROP TABLE are allowed to run concurrently.
-		return true
+		return action, true
+	case sqlparser.AlterDDLAction:
+		// ALTER is only allowed concurrent execution if this is a Vitess migration
+		strategy := onlineDDL.StrategySetting().Strategy
+		return action, (strategy == schema.DDLStrategyOnline || strategy == schema.DDLStrategyVitess)
 	case sqlparser.RevertDDLAction:
 		// REVERT is allowed to run concurrently.
-		// Reminder that REVERT is supported for CREATE, DROP and for 'online' ALTER, but never for
+		// Reminder that REVERT is supported for CREATE, DROP and for 'vitess' ALTER, but never for
 		// 'gh-ost' or 'pt-osc' ALTERs
-		return true
+		return action, true
 	}
-	return false
+	return action, false
 }
 
-// isAnyNonConcurrentMigrationRunning sees if there's any migration running right now
-// that does not have -allow-concurrent.
-// such a running migration will for example prevent a new non-concurrent migration from running.
-func (e *Executor) isAnyNonConcurrentMigrationRunning() bool {
-	nonConcurrentMigrationFound := false
-
-	e.ownedRunningMigrations.Range(func(_, val any) bool {
-		onlineDDL, ok := val.(*schema.OnlineDDL)
-		if !ok {
-			return true
-		}
-		if !e.allowConcurrentMigration(onlineDDL) {
-			// The migratoin may have declared itself to be --allow-concurrent, but our scheduler
-			// reserves the right to say "no, you're NOT in fact allowed to run concurrently"
-			// (as example, think a `gh-ost` ALTER migration that says --allow-concurrent)
-			nonConcurrentMigrationFound = true
-			return false // stop iteration, no need to review other migrations
-		}
+func (e *Executor) proposedMigrationConflictsWithRunningMigration(runningMigration, proposedMigration *schema.OnlineDDL) bool {
+	if runningMigration.Table == proposedMigration.Table {
+		// migrations operate on same table
 		return true
-	})
-
-	return nonConcurrentMigrationFound
+	}
+	_, isRunningMigrationAllowConcurrent := e.allowConcurrentMigration(runningMigration)
+	proposedMigrationAction, isProposedMigrationAllowConcurrent := e.allowConcurrentMigration(proposedMigration)
+	if !isRunningMigrationAllowConcurrent && !isProposedMigrationAllowConcurrent {
+		// neither allowed concurrently
+		return true
+	}
+	if proposedMigrationAction == sqlparser.AlterDDLAction {
```

Comment:
If I understand this correctly we allow only one copy phase for all online ddls across all tables. Is that right? If so, why can't we have multiple tables in the copy phase?

Reply:
> If I understand this correctly we allow only one copy phase for all online ddls across all tables. Is that right?

That is correct!

> If so, why can't we have multiple tables in the copy phase?

@rohit-nayak-ps I guess, the design of Online DDL is to have a distinct workflow per table. I'm not sure how feasible it would be to now aggregate multiple vreplication workflows into one, in the world of Online DDL.

Reply:
Are you saying that this is because Online DDL can use either gh-ost or vreplication? 

Didn't understand your point above of the need to aggregate multiple vreplication workflows into one: each table online ddl will have its own vreplication workflow and those can be in the copy phase concurrently, right? At least from a vreplication point of view?

Reply:
> Are you saying that this is because Online DDL can use either gh-ost or vreplication?

No, that wasn't what I meant ->

>  each table online ddl will have its own vreplication workflow and those can be in the copy phase concurrently, right

ah, right! Then, forget about what I wrote earlier. I misunderstood.

So, with multiple workflows, and single-table-per-workflow, there will also be multiple table locks and multiple `START TRANSACtioN wiTH CONSisteNT SNAPSHOT`, right? Which is what I want to avoid.
I fear running all copy phases together will put a substantial burden on the source database server. I don't have enough information to say whether running concurrent copy phases will be faster or slower than sequential copy phases; if it's significantly faster then that's an advantage we might consider; otherwise I feel it's better to lower the impact. There's a lot of "feel" here.

Reply:
That is a perfectly valid feeling, no issues with this decision! Just wanted to check if there was something fundamental that was stopping you in vreplication which we needed to address.

Reply:
Well it's better to have that feeling supported by data :stuck_out_tongue: